File: lichen/subsys/lichen/link/frame.c
/* SPDX-License-Identifier: GPL-3.0-or-later */
/* SPDX-FileCopyrightText: The contributors to the LICHEN project */

/**
 * @file frame.c
 * @brief LICHEN frame parsing and serialization
 */

#include <lichen/link.h>
#include <lichen/errno.h>
#include <string.h>

/* LLSec byte bit positions (spec/02-physical-link.md section 4.2) */
#define LLSEC_ADDR_MODE_MASK  0x03
#define LLSEC_MIC_LEN_SHIFT   2
#define LLSEC_MIC_LEN_MASK    0x1c
#define LLSEC_SIG_PRESENT     0x20
#define LLSEC_ENCRYPTED       0x40
#define LLSEC_SIID_PRESENT    0x80

/* Address lengths by mode (index = enum lichen_addr_mode value) */
static const uint8_t addr_lens[] = { 0, 2, 8, 0 };
#define ADDR_LENS_COUNT (sizeof(addr_lens) / sizeof(addr_lens[0]))

/* Compile-time assertions: ensure struct field sizes match max values */
_Static_assert(sizeof(((struct lichen_frame *)0)->dst_addr) >= 8,
	       "dst_addr must hold at least 8 bytes (EUI-64)");
_Static_assert(sizeof(((struct lichen_frame *)0)->mic) >= LICHEN_SIG_LEN,
	       "mic must hold a Schnorr-48 signature");

int lichen_frame_parse(struct lichen_frame *frame,
		       const uint8_t *data, size_t len)
{
	struct lichen_frame parsed = { 0 };

	if (frame == NULL || data == NULL) {
		return -EINVAL;
	}
	if (len > LICHEN_MAX_FRAME_LEN ||
	    (len > 0U && data[0] > LICHEN_MAX_FRAME_BODY_LEN)) {
		return -EMSGSIZE;
	}

	/*
	 * Minimum frame size: 5 bytes
	 *   length(1) + llsec(1) + epoch(1) + seqnum(2). Unsigned
	 * frames have no MIC; signed frames are checked after LLSec parsing.
	 */
	if (len < LICHEN_FRAME_FIXED_HEADER_LEN) {
		return -EINVAL;
	}

	size_t off = 0;
	uint8_t frame_len = data[off++];

	if (frame_len != len - 1) {
		return -EINVAL;
	}

	uint8_t llsec = data[off++];

	parsed.addr_mode = llsec & LLSEC_ADDR_MODE_MASK;

	/* SECURITY: Reject encrypted frames before any other LLSec policy so
	 * E=1 is always reported as unsupported encryption (spec 4.2). */
	parsed.encrypted = (llsec & LLSEC_ENCRYPTED) != 0;
	if (parsed.encrypted) {
		return -EPROTONOSUPPORT;
	}

	/* Only selectors 0b000 and 0b001 are defined; both mean no MIC on an
	 * unsigned frame. Selectors 0b010-0b111 are reserved. */
	uint8_t mic_length = (llsec & LLSEC_MIC_LEN_MASK) >> LLSEC_MIC_LEN_SHIFT;
	if (mic_length > LICHEN_MIC_64) {
		return -EINVAL;
	}

	parsed.mic_length = (enum lichen_mic_len)mic_length;
	parsed.signature_present = (llsec & LLSEC_SIG_PRESENT) != 0;

	/* SIID (bit 7) MUST equal the signature bit: signed frames set both
	 * and carry exactly one 8-byte signer EUI-64; unsigned frames clear
	 * both (spec/02-physical-link.md section 4.2). */
	parsed.signer_iid_present = (llsec & LLSEC_SIID_PRESENT) != 0;
	if (parsed.signer_iid_present != parsed.signature_present) {
		return -EINVAL;
	}
	parsed.signer_iid_len = parsed.signer_iid_present ? LICHEN_ADDR_MAX : 0U;

	/* Signed frames carry the full 48-byte Schnorr signature in the MIC
	 * field; unsigned frames have no MIC bytes regardless of selector. */
	parsed.mic_len = parsed.signature_present ? LICHEN_SIG_LEN : 0U;
	uint8_t addr_len = addr_lens[parsed.addr_mode];
	uint8_t signer_len = parsed.signer_iid_len;

	/* Check total required length: fixed header + address + signer + MIC */
	if (len < LICHEN_FRAME_FIXED_HEADER_LEN + (size_t)addr_len +
		   (size_t)signer_len + parsed.mic_len) {
		return -EINVAL;
	}

	parsed.epoch = data[off++];
	parsed.seqnum = (uint16_t)(((uint16_t)data[off] << 8) | data[off + 1]);
	off += 2;

	/* Destination address */
	parsed.dst_addr_len = addr_len;
	memcpy(parsed.dst_addr, &data[off], parsed.dst_addr_len);
	off += parsed.dst_addr_len;

	/* Signer EUI-64 (SIID), exactly when the SI bit is set */
	if (signer_len > 0) {
		memcpy(parsed.signer_iid, &data[off], signer_len);
		off += signer_len;
	}

	/* MIC at the end */
	memcpy(parsed.mic, &data[len - parsed.mic_len], parsed.mic_len);

	/* Payload is everything between address/signer IID and MIC */
	parsed.payload = &data[off];
	parsed.payload_len = len - off - parsed.mic_len;

	parsed.inner_payload_len = parsed.payload_len;

	/* Publish only a completely validated parse result. */
	*frame = parsed;

	return 0;
}

int lichen_frame_write(const struct lichen_frame *frame,
		       uint8_t *buf, size_t buflen)
{
	if (frame == NULL || buf == NULL) {
		return -EINVAL;
	}

	if ((unsigned int)frame->addr_mode >= ADDR_LENS_COUNT) {
		return -EINVAL;
	}

	if (frame->dst_addr_len != addr_lens[frame->addr_mode]) {
		return -EINVAL;
	}

	/* The MIC-length selector must be a defined value (0 or 1). */
	if ((unsigned int)frame->mic_length > LICHEN_MIC_64) {
		return -EINVAL;
	}

	/* SIID presence and length MUST match the signature flag: signed
	 * frames carry exactly one 8-byte signer EUI-64, unsigned frames
	 * carry none (spec/02-physical-link.md section 4.2). */
	if (frame->signer_iid_present != frame->signature_present ||
	    frame->signer_iid_len !=
		    (frame->signer_iid_present ? LICHEN_ADDR_MAX : 0U)) {
		return -EINVAL;
	}

	if (frame->encrypted) {
		return -EPROTONOSUPPORT;
	}

	/*
	 * Caller must initialize frame->mic and frame->mic_len before calling.
	 * The MIC is not computed here - it must be computed externally over
	 * the frame data and stored in frame->mic before serialization.
	 */
	uint8_t addr_len = addr_lens[frame->addr_mode];
	uint8_t signer_len = frame->signer_iid_len;
	uint8_t mic_len = frame->signature_present ? LICHEN_SIG_LEN : 0U;

	if (frame->mic_len != mic_len) {
		return -EINVAL;
	}

	/* Subtraction-first sizing: non_payload_len counts every byte of the
	 * serialized frame except the payload (including LENGTH), and the
	 * payload must fit in what remains of the 255-byte total. */
	size_t non_payload_len = LICHEN_FRAME_FIXED_HEADER_LEN + (size_t)addr_len +
				 (size_t)signer_len + mic_len;

	if (LICHEN_MAX_FRAME_LEN < non_payload_len ||
	    frame->payload_len > LICHEN_MAX_FRAME_LEN - non_payload_len) {
		return -EMSGSIZE;
	}
	if (frame->payload_len > 0U && frame->payload == NULL) {
		return -EINVAL;
	}

	size_t frame_len = non_payload_len + frame->payload_len;

	if (frame_len > buflen) {
		return -ENOMEM;
	}

	size_t off = 0;

	/* Length byte (excludes itself) */
	buf[off++] = (uint8_t)(frame_len - 1);

	/* LLSec byte — selector encoded from mic_length; signed frames set
	 * both the S and SI bits. */
	uint8_t llsec = frame->addr_mode & LLSEC_ADDR_MODE_MASK;
	llsec |= ((uint8_t)frame->mic_length << LLSEC_MIC_LEN_SHIFT) &
		 LLSEC_MIC_LEN_MASK;
	if (frame->signature_present) {
		llsec |= LLSEC_SIG_PRESENT | LLSEC_SIID_PRESENT;
	}
	buf[off++] = llsec;

	/* Epoch */
	buf[off++] = frame->epoch;

	/* Sequence number (big-endian) */
	buf[off++] = (uint8_t)(frame->seqnum >> 8);
	buf[off++] = (uint8_t)(frame->seqnum & 0xFF);

	/* Destination address */
	if (addr_len > 0) {
		memcpy(&buf[off], frame->dst_addr, addr_len);
	}
	off += addr_len;

	/* Signer EUI-64 */
	if (signer_len > 0) {
		memcpy(&buf[off], frame->signer_iid, signer_len);
	}
	off += signer_len;

	/* Payload */
	if (frame->payload_len > 0) {
		memcpy(&buf[off], frame->payload, frame->payload_len);
	}
	off += frame->payload_len;

	/* MIC */
	memcpy(&buf[off], frame->mic, mic_len);
	off += mic_len;

	return (int)off;
}

File: python/src/lichen/link/frame.py
# SPDX-License-Identifier: GPL-3.0-or-later
# SPDX-FileCopyrightText: The contributors to the LICHEN project
"""LICHEN link-layer frame format (spec section 4).

Wire layout (spec 4.1)::

    +--------+--------+-------+--------+----------+------+---------+--------+
    | Length | LLSec  | Epoch | SeqNum | Dst Addr | SIID | Payload | MIC    |
    +--------+--------+-------+--------+----------+------+---------+--------+
       1B       1B       1B      2B       0/2/8B    0/8B   var      0/48B

``Length`` is the total frame length excluding the Length field itself.
Multi-byte integer fields are big-endian.

The LLSec byte (spec 4.2) packs, from the least-significant bit::

    bits 0-1 : Addr Mode  (0=none/broadcast, 1=16-bit, 2=64-bit, 3=elided)
    bits 2-4 : MIC selector (0 or 1; ignored for wire MIC length)
    bit  5   : Signature present (Schnorr-48, 48 bytes)
    bit  6   : Encrypted (unsupported; receivers MUST reject)
    bit  7   : Signer identifier present (SI; canonical EUI-64; MUST equal S)
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum

# Exact application-domain prefix for every link Schnorr-48 transcript.
# Changing these octets is a protocol-version change.
LINK_SIGNATURE_DOMAIN = b"LICHEN-LINK-v1\x00"


class AddrMode(IntEnum):
    """Destination addressing mode (LLSec bits 0-1, spec 4.3)."""

    NONE = 0  # broadcast, 0 address bytes
    SHORT = 1  # 16-bit short address, 2 bytes
    EXTENDED = 2  # EUI-64, 8 bytes
    ELIDED = 3  # derived from IPv6 destination, 0 address bytes

    @property
    def addr_len(self) -> int:
        """Number of destination-address bytes for this mode."""
        return _ADDR_LEN_TABLE[self]


# Lookup table for AddrMode.addr_len (dict is robust against enum value changes).
_ADDR_LEN_TABLE: dict[AddrMode, int] = {
    AddrMode.NONE: 0,
    AddrMode.SHORT: 2,
    AddrMode.EXTENDED: 8,
    AddrMode.ELIDED: 0,
}


class MicLength(IntEnum):
    """MIC length / compatibility selector (LLSec bits 2-4, spec 4.2).

    Only 0 and 1 are valid (both mean no MIC on unsigned frames);
    values 2-7 are reserved. The enum is the source of truth for
    valid values (see from_bytes()).
    """

    BITS32 = 0  # compatibility selector; unsigned frames have no MIC
    BITS64 = 1  # compatibility selector; unsigned frames have no MIC

    def wire_mic_len(self, *, signature_present: bool) -> int:
        """Return the profile MIC width for this selector and signature state.

        The selector is retained for compatibility but does not select a wire
        width in the current profile. Unsigned frames carry no MIC, while both
        selector values carry the full Schnorr-48 value when signed.
        """
        if type(signature_present) is not bool:
            raise TypeError("signature_present must be bool")
        return _SIGNATURE_LENGTH if signature_present else 0


# LLSec bit fields.
_ADDR_MODE_MASK = 0b0000_0011
_MIC_LEN_SHIFT = 2
_MIC_LEN_MASK = 0b0000_0111
_SIGNATURE_BIT = 1 << 5
_ENCRYPTED_BIT = 1 << 6
_SI_BIT = 1 << 7

_MAX_FRAME_BODY = 254  # spec 4.1: Length field value is 4-254 bytes
MAX_FRAME_BODY = _MAX_FRAME_BODY
_SIGNATURE_LENGTH = 48  # Schnorr-48 signature


class FrameError(Exception):
    """Raised when a link-layer frame is malformed."""


class EncryptedFrameError(FrameError):
    """Raised when the unsupported encrypted-frame flag is present."""


@dataclass
class LichenFrame:
    """A parsed LICHEN link-layer frame.

    Attributes:
        epoch: 8-bit epoch counter (spec 4.4).
        seqnum: 16-bit sequence number (replay protection).
        dst_addr: Destination address bytes; length must match ``addr_mode``.
        payload: Frame payload (SCHC-compressed packet or app data).
        mic: MIC or Schnorr-48 signature (48 bytes if signature_present=True,
            else per mic_length; current profile uses signature in MIC field).
        addr_mode: Destination addressing mode.
        mic_length: MIC length selector (ignored for length when signed).
        signature_present: Whether Schnorr-48 signature is present in MIC field
            (LLSec bit 5; see draft-lichen-schnorr-00). The signature lives
            entirely in the 48-byte MIC field; the payload is delivered to
            consumers whole, with no bytes stripped for signatures.
        encrypted: Whether the unsupported encrypted-frame flag is set.
        signer_eui64: The signer's canonical EUI-64 wire value. It is modified
            into an IPv6 IID only when resolving the link address. Present
            exactly when signed.
    """

    epoch: int
    seqnum: int
    dst_addr: bytes
    payload: bytes
    mic: bytes
    addr_mode: AddrMode = AddrMode.NONE
    mic_length: MicLength = MicLength.BITS32
    signature_present: bool = False
    encrypted: bool = False
    signer_eui64: bytes = b""

    def _validate(self) -> None:
        if type(self.addr_mode) is not AddrMode:
            raise FrameError("addr_mode must be an AddrMode")
        if type(self.mic_length) is not MicLength:
            raise FrameError("mic_length must be a supported MicLength selector")
        if type(self.signature_present) is not bool:
            raise FrameError("signature_present must be bool")
        if type(self.encrypted) is not bool:
            raise FrameError("encrypted must be bool")
        if type(self.epoch) is not int:
            raise FrameError("epoch must be an integer")
        if type(self.seqnum) is not int:
            raise FrameError("seqnum must be an integer")
        for name, value in (
            ("dst_addr", self.dst_addr),
            ("payload", self.payload),
            ("mic", self.mic),
            ("signer_eui64", self.signer_eui64),
        ):
            if type(value) is not bytes:
                raise FrameError(f"{name} must be bytes")
        if self.encrypted:
            raise FrameError("encrypted frames are unsupported")
        if not 0 <= self.epoch <= 0xFF:
            raise FrameError(f"epoch out of range: {self.epoch}")
        if not 0 <= self.seqnum <= 0xFFFF:
            raise FrameError(f"seqnum out of range: {self.seqnum}")
        if len(self.dst_addr) != self.addr_mode.addr_len:
            raise FrameError(
                f"dst_addr is {len(self.dst_addr)} bytes but {self.addr_mode.name} "
                f"requires {self.addr_mode.addr_len}"
            )
        expected_mic_len = _SIGNATURE_LENGTH if self.signature_present else 0
        if len(self.mic) != expected_mic_len:
            raise FrameError(f"mic is {len(self.mic)} bytes but {expected_mic_len} are required")
        if bool(self.signer_eui64) != self.signature_present:
            raise FrameError("signature and signer EUI-64 presence must match")
        if self.signature_present and len(self.signer_eui64) != 8:
            raise FrameError("signed frame requires an 8-byte signer EUI-64")

    def llsec_byte(self) -> int:
        """Compute the LLSec flags byte."""
        value = int(self.addr_mode) & _ADDR_MODE_MASK
        value |= (int(self.mic_length) & _MIC_LEN_MASK) << _MIC_LEN_SHIFT
        if self.signature_present:
            value |= _SIGNATURE_BIT
        if self.encrypted:
            value |= _ENCRYPTED_BIT
        if self.signer_eui64:
            value |= _SI_BIT
        return value

    def to_bytes(self) -> bytes:
        """Serialize the frame to its on-air byte representation.

        Raises:
            FrameError: If a field is out of range, lengths are inconsistent
                with the LLSec modes, or the frame exceeds 254 body bytes.
        """
        self._validate()
        body_len = (
            4 + len(self.dst_addr) + len(self.signer_eui64) + len(self.payload) + len(self.mic)
        )
        if body_len > MAX_FRAME_BODY:
            raise FrameError(f"frame body is {body_len} bytes, exceeds {MAX_FRAME_BODY}")
        body = (
            bytes([self.llsec_byte(), self.epoch])
            + self.seqnum.to_bytes(2, "big")
            + self.dst_addr
            + self.signer_eui64
            + self.payload
            + self.mic
        )
        return bytes([len(body)]) + body

    @classmethod
    def from_bytes(cls, data: bytes) -> LichenFrame:
        """Parse a frame from its on-air byte representation.

        Raises:
            FrameError: If the data is truncated, the length field is wrong, the
                signature/SIID flags disagree, or the MIC-length field is reserved.
        """
        if type(data) is not bytes:
            raise FrameError("frame must be bytes")
        if len(data) < 1:
            raise FrameError("frame is empty")
        if len(data) > MAX_FRAME_BODY + 1:
            raise FrameError(f"frame is {len(data)} bytes, exceeds {MAX_FRAME_BODY + 1}")
        length = data[0]
        if length > MAX_FRAME_BODY:
            raise FrameError(f"frame body is {length} bytes, exceeds {MAX_FRAME_BODY}")
        received_body_len = len(data) - 1
        if received_body_len != length:
            raise FrameError(
                f"length field says {length} but {received_body_len} body bytes present"
            )
        body = data[1:]
        # Fixed fields: LLSec(1) + Epoch(1) + SeqNum(2) = 4 bytes minimum.
        if length < 4:
            raise FrameError(f"frame body too short: {length} bytes")

        llsec = body[0]
        # SECURITY: Encrypted frames are unsupported; receivers MUST reject
        # them before signature or reserved-bit processing so E=1 always
        # reports as unsupported encryption (spec 4.2).
        if llsec & _ENCRYPTED_BIT:
            raise EncryptedFrameError("encrypted frames are unsupported")
        addr_mode = AddrMode(llsec & _ADDR_MODE_MASK)
        mic_field = (llsec >> _MIC_LEN_SHIFT) & _MIC_LEN_MASK
        try:
            mic_length = MicLength(mic_field)
        except ValueError:
            # SECURITY: A malicious frame could set the signature bit (bit 5)
            # while using a reserved MIC-length value to claim no signature
            # bytes follow, causing the receiver to parse signature bytes as
            # payload.  Rejecting reserved values closes this vector.
            raise FrameError(f"reserved MIC-length value: {mic_field}") from None

        epoch = body[1]
        seqnum = int.from_bytes(body[2:4], "big")

        offset = 4
        addr_len = addr_mode.addr_len
        signature_present = bool(llsec & _SIGNATURE_BIT)
        signer_eui64_present = bool(llsec & _SI_BIT)
        if signer_eui64_present != signature_present:
            raise FrameError("signature and signer EUI-64 presence bits must match")
        # SECURITY: Reject frames where signature_present is set but the frame
        # body is too short for the 48-byte Schnorr signature. An attacker could
        # set the signature bit without appending a signature, hoping the parser
        # reads past the buffer (over-read) or misinterprets payload bytes as MIC.
        mic_len = _SIGNATURE_LENGTH if signature_present else 0
        signer_eui64_len = 8 if signer_eui64_present else 0
        if length < offset + addr_len + signer_eui64_len + mic_len:
            raise FrameError("frame too short for declared address/MIC sizes")

        dst_addr = body[offset : offset + addr_len]
        offset += addr_len
        signer_eui64 = body[offset : offset + signer_eui64_len]
        offset += signer_eui64_len
        payload = body[offset : len(body) - mic_len]
        mic = body[len(body) - mic_len :]

        return cls(
            epoch=epoch,
            seqnum=seqnum,
            dst_addr=dst_addr,
            payload=payload,
            mic=mic,
            addr_mode=addr_mode,
            mic_length=mic_length,
            signature_present=signature_present,
            encrypted=bool(llsec & _ENCRYPTED_BIT),
            signer_eui64=signer_eui64,
        )
